### ISSUE TITLE
Max concurrent per project for CPU thread count

### ISSUE THREAD

**Issue by hucker75** [BOINC/boinc#5046]
I currently use:

<app_config>
   <project_max_concurrent>N</project_max_concurrent>
</app_config>

on the LHC project, because running too many of their tasks (all VB) slows my PC to a halt.  This works fine until an ATLAS task appears.  They use 8 cores, while the CMS and Theory use 1.  So I get all the cores used at once and the system overloads.  Could we have a similar thing for max concurrent threads per project?

**Comment by computezrmle** [comment]
Either use the prefs page at LHC@home to limit the #cores for all your computers connected to the same venue or use an app_config.xml to limit the #cores for an app running under an individual BOINC client instance.

**Comment by hucker75** [comment]
If you mean setting "Max # of CPUs for this project" in the venue, that only allows up to 8.  Does that mean number of threads?  If so most computers have more threads than that, the one I have a problem with has 24.

I don't want to limit per app, I want to limit the total cores for all running VB tasks.

P.S. Atlas on the defaults only makes use of 2 cores out of the 8, it used to use all 8.   Have they changed it or is that my machine being odd?  I changed it in app config to only use 2, and it still used 2, but still used 10GB, which limits things somewhat.  I checked some of your jobs, and I see your Atlas tasks are also 2.

**Comment by computezrmle** [comment]
To me it looks like the issue is based on project limits instead of general BOINC limits.


As for "Max # of CPUs for this project" LHC@home allows values between 1-8 plus unlimited.
This is a per task setting, but the project server is configured not to send more than 2 * #cores = max_tasks to a host.

If unlimited is set by the volunteer, vbox apps are limited to 8 cores, ATLAS native (Linux only) is limited to 12 cores.
Computers with less cores are limited to their #cores.

The #cores is used to calculate the amount of RAM for vbox apps.
See the project forum for details.

#cores and RAM value can be modified by the volunteer using an app_config.xml.


> I checked some of your jobs ...

Doesn't make sense since your Windows computers run vbox apps while my Linux computers run native apps.
In addition I'm using an app_config.xml that configures values optimized for my project mix.
It's very likely that your computers require other settings.

**Comment by hucker75** [comment]
I have my LHC settings on unlimited.  I do get 8 core atlas tasks, but they never use more than 2 cores.  Months ago when I last ran Atlas, they used all 8.

Are you saying if I set max cores to 4, I would get no more than 8 tasks per host?  But if they're Atlas that would be more cores than if they were CMS.  So that doesn't help.

So, is there a way to simply say I want to use 16 of my 24 cores on these machines?

**Comment by Toby-Broom** [comment]
I agree the settings are not easy to understand, I gave up commenting about it.

I have No Limit on the site for both, and then set the the rest to be:

`<app_config>

<app_version> 
	<app_name>ATLAS</app_name> 
	<plan_class>vbox64_mt_mcore_atlas</plan_class> 
	<avg_ncpus>1.0</avg_ncpus> 
	<cmdline>--nthreads 1 --memory_size_mb 10200</cmdline>
</app_version>
</app_config>`

They use tons of RAM but I did it like this becuase of the #4882 if it set it to lower then it doesn't actually start more tasks since it thinks its using 10 GB anyhow.

**Comment by hucker75** [comment]
The trouble with that is if it thinks it's used all the RAM, other projects won't run either.

**Comment by Toby-Broom** [comment]
Yes, it acutally uses 10 GB with this setup, if you run Linux then you can use use whatever you like for the --memory_size_mb and it works as expected.  the memory issue is just on windows.

**Comment by hucker75** [comment]
I don't follow.  If you're limiting how many LHC tasks run by memory settings, aren't you also stopping another project using your remaining cores?

**Comment by Toby-Broom** [comment]
If there is not enough RAM,.. only ATLAS uses 10 GB so if you have a mix of WUs then its not so bad. if you add 
<app>
      <name>ATLAS</name>
      <max_concurrent>5</max_concurrent>
    </app>

Then it would only ever use 50 GB for ATLAS leaving free for other programs, that what I did when I only had 128 GB of ram on windows. 

You could also say set --nthreads 2, then each WU will use 2 cores and 10 GB so you half your ram and double the cores.  Since were editing just the ATLAS WU's then CMS etc are not efffected

**Comment by hucker75** [comment]
The trouble is I wanted to limit total cores for LHC, as VB programs slow the interface.  For example I'd be happy with one 8-core Atlas and four 1-core CMS, or.... twelve of 1-core CMS.  But I can't set that.

**Comment by Toby-Broom** [comment]
What about if you set ATLAS to --nthreads 1 use all to one and use projectmax concurent. the ATLAS will still use 10 GB unless you set the Max # CPUs = 1, I think there is a limt to the number of queued WU = n threads on your CPU. 

The websetting are kind of funky in how they work, you might need to set max jobs to 1 and set the max cpu to unlimited.
